### PR TITLE
Ensure lazy evaluation for probs and logits

### PR DESCRIPTION
**Pull request by neerajprad**
This makes a few modifications to ensure that `logits` and `probs` are not accessed by a method unless it needs to. e.g. accessing `probs` on a class that was initialized with `logits` only to get the batch shape or event shape. This affects Categorical, OneHotCategorical, Bernoulli, Binomial and Multinomial. 
 - Added `.new()` methods to these distributions that can help with generating tensors/variables of the right type and placing it on the correct CUDA device.

**Testing:** Added some tests to prevent future regression:
 - `initializing` a distribution with `probs` and calling `sample`, `enumerate_support`, `batch_shape` and `event_shape` should not initialize `logits`.
 - Conversely, `initializing` a distribution with `logits` and calling `log_prob`, `enumerate_support`, `batch_shape` and `event_shape` should not initialize `probs`.

### CODE REVIEW

**Comment by neerajprad** [comment]
cc. @apaszke, @alicanb, @fritzo.

File: torch/distributions/multinomial.py [hunk 1]
```diff
@@ -73,7 +74,7 @@ def sample(self, sample_shape=torch.Size()):
         samples = samples.permute(*shifted_idx)
         counts = samples.new(self._extended_shape(sample_shape)).zero_()
         counts.scatter_add_(-1, samples, torch.ones_like(samples))
-        return counts.type_as(self.probs)
+        return counts.type_as(self._param)
```
Comment (fritzo) [inline torch/distributions/multinomial.py]: (not part of this PR, but) Does `.type_as()` get GPU placement right? Or should we instead be using `self._param.new(...)`?

**Review by fritzo — COMMENTED** [review]
Thanks for doing this!

File: test/test_distributions.py [hunk 2]
```diff
@@ -1732,5 +1732,47 @@ def test_multinomial_log_prob_with_logits(self):
...
+                dist.sample()
+                message = 'Failed for {} example 0/{}'.format(Dist.__name__, len(params))
+                self.assertFalse('logits' in vars(dist), msg=message)
+                try:
+                    dist.enumerate_support()
+                except NotImplementedError:
+                    pass
+                self.assertFalse('logits' in vars(dist), msg=message)
+                batch_shape, event_shape = dist.batch_shape, dist.event_shape
+                print vars(dist)
```
Comment (alicanb) [inline test/test_distributions.py]: python3 problem here

Comment (neerajprad) [reply test/test_distributions.py]: My bad..missed removing this.

**Comment by alicanb** [comment]
BTW @apaszke  just merged binomial PR so can we change binomial in this PR as well?

**Comment by neerajprad** [comment]
> BTW @apaszke just merged binomial PR so can we change binomial in this PR as well?

Sure will address that in this PR.

**Comment by neerajprad** [comment]
@fritzo - Continuing the thread from https://github.com/probtorch/pytorch/pull/85, do we want to add `.new` methods to the distributions directly, or just use `param.new()` as needed? The former seems a bit ambiguous to me since the parameters' shape might be different from the distribution's sample shape.

**Comment by fritzo** [comment]
I believe a sufficient API would be a `.new()` method and a `.param_shape` property. The current code uses a few other methods and attributes like `.is_cuda` but I believe `.new()` and `.param_shape` should suffice. It seems preferable to me to expose `Dist.new()` rather than `Dist.param.new()` to avoid giving the impression that `.param` was in any way a complete parameterization.

**Comment by fritzo** [comment]
@vishwakftw If we implement `.new()` methods, then I believe we can implement a single
```py
def _kl_infinite(p, q):
    return p.new([float('inf')])
```

File: torch/distributions/binomial.py [hunk 3]
```diff
@@ -64,25 +69,30 @@ def logits(self):
...
     def log_prob(self, value):
         self._validate_log_prob_arg(value)
-        probs = clamp_probs(self.probs)
         log_factorial_n = math.lgamma(self.total_count + 1)
         log_factorial_k = torch.lgamma(value + 1)
         log_factorial_nmk = torch.lgamma(self.total_count - value + 1)
+        max_val = (-self.logits).clamp(min=0.0)
         return (log_factorial_n - log_factorial_k - log_factorial_nmk +
-                value * self.logits + self.total_count * torch.log1p(-probs))
+                value * self.logits + self.total_count * max_val -
```
Comment (neerajprad) [inline torch/distributions/binomial.py]: @alicanb - I wanted to eliminate the use of probs. Does this look fine to you?

File: torch/distributions/one_hot_categorical.py [hunk 4]
```diff
@@ -53,11 +60,13 @@ def entropy(self):
         return self._categorical.entropy()
 
     def enumerate_support(self):
-        probs = self._categorical.probs
         n = self.event_shape[0]
-        if isinstance(probs, Variable):
-            values = Variable(torch.eye(n, out=probs.data.new(n, n)))
+        out = self._categorical.new((n, n))
+        if isinstance(out, Variable):
```
Comment (neerajprad) [inline torch/distributions/one_hot_categorical.py]: @fritzo - let me know if there is a better way to handle this.

Comment (fritzo) [reply torch/distributions/one_hot_categorical.py]: I guess you could simplify a tiny bit
```py
values = self._categorical.new((n, n))
torch.eye(n, out=(values.data if isinstance(values, Variable) else values)
```
If `torch.eye(out=values)` really doesn't work for CUDA then we should ask Adam what to do.

File: torch/distributions/binomial.py (hunk 3 above)
Comment (alicanb) [reply torch/distributions/binomial.py]: Sure, we can get rid of `clamp_probs` then, I think this was the only place that it's used

File: torch/distributions/bernoulli.py [hunk 5]
```diff
@@ -34,16 +34,21 @@ def __init__(self, probs=None, logits=None):
...
+        self._param = self.probs if probs is not None else self.logits
+        if is_scalar:
             batch_shape = torch.Size()
         else:
-            batch_shape = probs_or_logits.size()
+            batch_shape = self._param.size()
         super(Bernoulli, self).__init__(batch_shape)
 
+    def new(self, *args, **kwargs):
+        return self._param.new(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply torch/distributions/bernoulli.py]: Will change to `_new` instead; agreed that it does not have to be part of the public interface.

[the remark this answers is not in the document]
Comment (fritzo) [reply torch/distributions/bernoulli.py]: Sorry for the confusion.

**Review by apaszke — APPROVED** [review]
